File: typescript/src/envelopes.ts
export interface SnapshotEnvelope<Stream, Payload> {
  readonly kind: 'snapshot';
  readonly stream: Stream;
  readonly sequence: number;
  readonly payload: Payload;
}

export interface DeltaEnvelope<Stream, Payload> {
  readonly kind: 'delta';
  readonly stream: Stream;
  readonly sequence: number;
  readonly payload: Payload;
}

export interface CommandEnvelope<Command, Payload> {
  readonly commandId: Command;
  readonly expectedVersion?: number;
  readonly payload: Payload;
}

export type CommandReceipt<Command, Result, Rejection = string> =
  | {
      readonly commandId: Command;
      readonly status: 'accepted';
      readonly result: Result;
    }
  | {
      readonly commandId: Command;
      readonly status: 'rejected';
      readonly rejection: Rejection;
    };

File: rust/src/envelopes.rs
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct SnapshotEnvelope<Stream, Payload> {
    pub stream: Stream,
    pub sequence: u64,
    pub payload: Payload,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct DeltaEnvelope<Stream, Payload> {
    pub stream: Stream,
    pub sequence: u64,
    pub payload: Payload,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct CommandEnvelope<Command, Payload> {
    pub command_id: Command,
    pub expected_version: Option<u64>,
    pub payload: Payload,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub enum CommandReceipt<Command, Result, Rejection = String> {
    Accepted {
        command_id: Command,
        result: Result,
    },
    Rejected {
        command_id: Command,
        rejection: Rejection,
    },
}
